### ISSUE TITLE
Support multidimensional kernel launches

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#22152]
## Background

This feature request has come up in different contexts.

#### 1. Matrix transpose

For team members at HPE this was in https://github.com/Cray/chapel-private/issues/4288

@DanilaFe requested this with:

> The CUDA [transpose benchmark](https://www.nvidia.com/content/cudazone/cuda_sdk/Linear_Algebra.html#transpose), as well as other transpose implementations we've seen, uses a two-dimensional kernel launch to implement the read-write coalesced version. This is because it is desirable to make reads from the input array and writes to the output array occur row-wise (i.e., not strided). To this end, the coalesced implementation creates a 16x16 memory segment in block shared memory, reads to it row-wise (writing columns) and writes from it row-wise (reading rows). The two-dimensional block size is useful for creating and populating this square piece of shared memory.
> 
> Currently, to implement such an algorithm in Chapel requires using a #pragma "codegen for gpu" proc, and the multi-dimensional kernel launch primitive. That is to say, there is no user-facing way to launch a multi-dimensional kernel. However, judging from this transpose benchmark, it seems like we might need to consider ways for users to write code that results in multi-dimensional blocks. We want it to be possible to write a performant transpose using idiomatic Chapel.
> 
> There are some open questions, including the behavior of multi-dimensional zippering (as well as the precise sort of language features that might be necessary to have finer-grained control over block shapes, and even how loops over two-dimensional domains should be GPUized).

@mppf's thoughts were:

> My understanding is that other languages / efforts with GPUs have run into the problem that the desired traversal order on GPU to achieve the best memory coalescing is different from the best traversal order on the CPU to achieve the best cache usage. This might apply to row-major vs. column-major but at least I have heard of it applying to array-of-structures vs structure-of-arrays. It is unclear to me to what degree we have this problem, but when thinking about multidimensional loops and zippering, I think a key thing to consider is what the most natural memory access pattern will be & whether or not we can make that efficient on both the GPU and the CPU.

#### 2. @psath alluded to this need

See https://chapel.discourse.group/t/mixing-atomic-add-with-non-atomic-read-and-towards-gpu-atomics/21350

#### Standalone functions now, new features later

How can we support this? I am somewhat strongly for using something similar to `setBlockSize` today:

```chapel
on here.gpus[0] {
  foreach i in 1..n {
    setGridSize(x,y,z);  // would make the grid 3 dimensional
    setBlockSize(a,b,c);  // would make the block 3 dimensional
  }
}
```

`setBlockSize` is a temporary solution, though. I expect that we replace these with more portable solutions via something like https://github.com/chapel-lang/chapel/issues/16405#issuecomment-991253890. Through that a potential snippet might look like:

```chapel
on here.gpus[0] {
  foreach i in 1..n with (config cfg = new LoopContext(gridSize=(x,y,z), blockSize=(x,y,z))) {

  }
}
```

Note that this snippet also relies on `foreach` intents, which are not supported but requested [here](https://github.com/chapel-lang/chapel/issues/18500).

#### Nested loops

An alternative is to compile nested order-independent loops into multidimensional kernels:

```chapel
on here.gpus[0] {
  forall i in gridDimX {
    forall j in gridDimY {
      forall k in gridDimZ {
      }
    }
  }
}
```

Implementation challenges aside, I think this is nested parallelism is more akin to a kernel launching another kernel aka dynamic parallelism rather than multidimensional kernels. I'll create an issue for dynamic parallelism, too.
 
## Implementation Details

I'll ~~ramble a bit about~~ capture my thoughts here:

Implementing this would involve translating the loop indices into thread indices. It may be easy for 1D loops, but more challenging for multidimensional loops, which I expect to be a more common case in scenarios where you'd want a multidimensional kernel. Today, for multidimensional loops, the outermost dimension "becomes" the kernel, where the inner dimensions are regular for loops within. First, we'll need to change that, which is probably the main challenge, but I don't think it is an obstacle. Second, we'll need to translate the thread indices into loop indices somehow. There are questions around whether you can have a 3D loop with a 2D kernel launch for example. If we were to support that, we'll need to make the outermost 2 dims the kernel, and the innermost dimension to be a for loop within. And maybe we can prevent number of ranks of the kernel from being greater than the number of ranks of the loop as the initial step? @stonea I believe you implemented support for multidimensional loops as we have today. What are your thoughts here?

**Comment by psath** [comment]
I'd initially been leaning towards the nested forall, albeit inversed so the fastest-varying/X is innermost.

But after looking at your second example, and the description of zippered iterators, that seems like a natural 4th option.
> zippered iteration causes the loop body to be executed once per corresponding set of values yielded by the iterands.

(which sounds a lot like `(threadIdx.x + blockDim.x*blockIdx.x)` or OpenCL's `get_global_id()`

i.e. restrict the zipper to 3D somehow and then do something like this:
```
on here.gpus[0] {
  forall (x, y, z) in zip(1..gridDim.x*blockDim.x, 1..gridDim.y*blockDim.y, 1..gridDim.z*blockDim.y) {
    ...
  }
}
```
In practice I'd still want to know which dimension was fastest-varying, to coordinate warps/wavefronts and shared memory access, but that's just documentation after the fact.

Which would leave the nested forall for either dynamic parallelism or asynchronous dispatch. (Probably more the former than the latter.)

**Comment by stonea** [comment]
> @stonea I believe you implemented support for multidimensional loops as we have today. What are your thoughts here?

I'm not sure to what extent that was me and to what extent it was @DanilaFe.  But I'll give you my initial overall impressions ---

Just looking at your examples, it seems a lot less ambiguous what should occur with the LoopContext approach then with the nested loop approach. That said, the nested loops thing certainly seems like something users would be naturally inclined to expect to work.

With the LoopContext approach:
  - I think we could get away without specifying grid size explicitly, that should just follow from the loop indices and block size right?
  - So far we've been assuming a one-to-one mapping between threads and iterations but maybe we should allow multiple iterations per thread, perhaps we could have another parameter that amounts to the "block size" of indices covered by a single thread?
  - I think we would want some mode that amounts to "hey Chapel figure this out for me", which would then call into `cudaOccupancyMaxPotentialBlockSize` or something similar (or maybe this is what we should just do if users don't explicitly specify something).

And about the nested loop approach:
  - If we go this route, I'm wondering if we would still want to support some way of having the old approach where the outer loop GPUized and the inner loops were turned into sequential loops.
    - I supose we could just tell users to manually use for loops for the inner loops if that's what the want but maybe there's some reason they wouldn't want to? What would the behavior be in regards to vectorization on the CPU with nested foralls/foreaches?
  - CUDA/HIP only support 3 dimensions for block/grid size so after the third loop I guess we'd keep our existing behavior of having it turn to for loops?
  - Block size and grid size needs to be known at kernel-launch time so we would need to calculate the inner loop bounds before the outer one starts executing.  In some cases this is doable but in others it isn't: what if the inner loop bound makes use of the index variable of the outer one, or what if it calls a non-pure function?

For this last point, consider the following loop for example. Say we computed the loop bounds before doing a kernel launch, in this case if N=0 we'd run into a divide by zero error, so we'd have to add some extra checks to handle cases like this as well:

```chpl
config const N;
forall i in 0..N {
  forall j in 0..(100/N) {
    ...
  }
}
```

It also occurs to me that maybe this doesn't have to be an either/or type thing.

Maybe we could:
  - support the nested forall/foreach case for the simple case where the loop bounds can be precomputed, users want a one-to-one mapping of iterations to threads, and they'd be fine with Chapel choosing a block size on their behalf
  - allow overriding this behavior using loop config approach for when users want to intervene
  - in mixed cases (where there's loop configs and nested loops) we have the loops that don't explicitly have a loop configuration fallback to the old behavior of turning into normal for loops.

**Comment by stonea** [comment]
I've created a new GH issue about having a non one-to-one mapping here: https://github.com/chapel-lang/chapel/issues/22175

**Comment by e-kayrakli** [comment]
> By clauses currently prevent GPUization, but IIRC yall already have an issue about that?

Yes, if the stride is non-param. We didn't, but it was something we were aware of. Thanks for bringing it up. I opened:

- https://github.com/chapel-lang/chapel/issues/22176 that closely captures what you're referring to as a sidenote
- https://github.com/chapel-lang/chapel/issues/22177 is the underlying GPU limitation for the problem
- Spiraling out; https://github.com/chapel-lang/chapel/issues/22178 has a potential step towards fixing #22177
